### ISSUE TITLE
Reproducing inference without CRF

### ISSUE THREAD

**Issue by guillembraso** [muyangyi/SimSeg#11]
Hi, 

Thanks again for your work!
I have question regarding running inference without CRF. In Table 4 of your paper, you report a 2-3 mIoU point decrease, when disabling the CRF. Could you please provide your inference code disabling CRF?

I have tried the following:

```       
min_value, max_value = attn_ai2at.min(), attn_ai2at.max()
norm_attn = (attn_ai2at - min_value) / (max_value - min_value)

### use crf
if not disable_crf:
    binary_mask = dense_crf(np.array(image_a).astype(np.uint8), norm_attn) * 255
    binary_mask = binary_mask.astype('uint8')

else:
    binary_mask = (norm_attn * 255).astype('uint8')
```

in line `153` of `tools/seg_evaluation.py`. And it yields a dramatic drop to `~18` mIoU. I have tried a couple of alternatives, e.g. not using unnormalized feature maps, or not multiplying by image-level scores (as done [here](https://github.com/muyangyi/SimSeg/blob/025f8399f8930fdbf5079d33ebb3ebeeb71ec308/tools/seg_evaluation.py#L162)), and results look similarly bad.

It'd be great if you could help me reproduce your numbers without CRF. 

Thanks!

Guillem

**Comment by muyangyi** [comment]
```
def evaluate_benchmark(loader, model, cfg, tokenizer, seg_categories, top_cls_num):
    image_mean = torch.tensor(cfg.transforms.normalize.mean, device=ENV.device)
    image_mean = image_mean.view(1, 3, 1, 1)
    image_norm = torch.tensor(cfg.transforms.normalize.std, device=ENV.device)
    image_norm = image_norm.view(1, 3, 1, 1)

    patch_size = 16
    num_patch = cfg.transforms.input_size // patch_size

    # extract text features for labels
    label_text_feature = zero_shot_classifier(model, seg_categories, openai_imagenet_template, tokenizer, ENV)

    count = 0
    total_intersection, total_union = 0, 0

    for batch in tqdm(loader):
        batch_dict = {}
        batch_dict['image'], batch_dict['mask_label'] = batch
        batch_dict = {k: v.cuda(ENV.device, non_blocking=True) for k,v in batch_dict.items() if k not in ['caption']}

        # image_feature includes [CLS] token
        image_feature = model.module.forward_image_feature(batch_dict['image']) # (Batch, 324, 768)
        # projector and L2 normalize
        image_feature_pooled = model.module.forward_image_project(image_feature) # (Batch, 512)
        image_feature = model.module.image_projection(image_feature) # (Batch, 324, 512)

        # visualize raw image and text
        image_raw = (((batch_dict['image'] * image_norm) + image_mean) * 255).type(torch.uint8).permute(0, 2, 3, 1).cpu().numpy()
        image_shape = batch_dict['mask_label'][0].shape
        mask_label = batch_dict['mask_label']

        index = 0
        image_a = Image.fromarray(image_raw[index])
        im_f_a = image_feature[index]
        im_f_a = F.normalize(im_f_a, dim=-1, p=2)
        im_avg_a = image_feature_pooled[index]

        attn_ai2ai = 1.0 * im_f_a @ im_avg_a.unsqueeze(-1) # (324, 1)
        attn_ai2ai = attn_ai2ai.reshape(num_patch, num_patch)
        attn_ai2ai = F.interpolate(attn_ai2ai.unsqueeze(0).unsqueeze(0), scale_factor=patch_size, mode="nearest")[0][0]

        i2t_scores = torch.sum(im_avg_a.unsqueeze(0) * label_text_feature, dim=1)
        scores = i2t_scores 

        topk_scores, topk_index = scores.topk(top_cls_num)
        score_mean, score_std = topk_scores.mean(), topk_scores.std()
        threshold = score_mean + 1.5 * score_std

        raw_H, raw_W = image_shape
        temp_pred = np.zeros((len(seg_categories), raw_H, raw_W))

        candidate_class_num = 5

        for i, index in enumerate(topk_index[:candidate_class_num]):
            
            if index in [0,255]:
                continue

            attn_ai2at = im_f_a @ label_text_feature[index].unsqueeze(-1) # (324, 1)
            attn_ai2at = attn_ai2at.reshape(num_patch, num_patch)

            attn_ai2at = F.interpolate(attn_ai2at.unsqueeze(0).unsqueeze(0), scale_factor=patch_size, mode="bilinear")[0][0] 

            H, W = attn_ai2at.shape[-2:]
            attn_ai2at = attn_ai2at.reshape(H*W, 1)
            attn_ai2at = torch.softmax(1.0 * attn_ai2at, dim=0)
            attn_ai2at = attn_ai2at.reshape(H, W).cpu().numpy()

            score = float(scores[index])
            if score < threshold:
                break

            min_value, max_value = attn_ai2at.min(), attn_ai2at.max()
            norm_attn = (attn_ai2at - min_value) / (max_value - min_value)

            # no crf
            final_mask = (norm_attn > 0.5) * 255

            final_mask = cv2.resize(final_mask.astype(np.uint8), dsize=(raw_W, raw_H), interpolation=cv2.INTER_NEAREST)
            temp_pred[index] = final_mask * score
```

**Comment by muyangyi** [comment]
There are 4 places being changed:
-  `threshold = score_mean + 1.5 * score_std`, 1.0 -> 1.5
-  `attn_ai2at = F.interpolate(attn_ai2at.unsqueeze(0).unsqueeze(0), scale_factor=patch_size, mode="bilinear")[0][0]`, nearest -> bilinear
-  `attn_ai2at = torch.softmax(1.0 * attn_ai2at, dim=0)`, newly add
-  `final_mask = (norm_attn > 0.5) * 255` replace the crf & dilate & erode

**Comment by guillembraso** [comment]
Hey! forgot to follow-up on this. Thanks for your response. With the code you provided I was able to reproduce the results you report on the paper on VOC and context. However, for COCO, I am getting close to 0 mIoU. Are there any hyperparameters that require adaptation there?

**Comment by muyangyi** [comment]
For COCO, we don't have specific hyperparameters.    
However, we only calculate mIoU on 81 objects (not 172) as listed in `data/label_category/coco_stuff.txt`.
